### ISSUE TITLE
Figure out how to deal with cmaes's es.stop() better

### ISSUE THREAD

**Issue by MichaelClerx** [pints-team/pints#577]
CMA-ES, has a long list of built-in stopping criteria. For consistent behaviour between solvers, we want to override these with our own criteria.

At the moment, we handle by asking CMA-ES if we need to stop on every iteration, but ignoring what it says unless it's an error message (in which case we halt the routine via `Optimisation.stop()`).

I've just added in some code that logs ignored CMA-ES stopping attempts, and for the logistic example it wants to stop quite a bit before we do (about 20 iterations too early, in fact, but a few hundred iterations before we stop (with no discernible improvement during that time, so probably a good thing).

Stopping early is good! So how do we include this in Pints?

Ideas:

1. Add some option to say "stop when CMA-ES wants to, I don't care if it's inconsistent across optimisers". This would mean we'd have to provide the user with access to the underlying cma object to set the different stopping criteria, at which point they might as well skip Pints and just use `cma`.
2. Get to know CMA-ES stopping criteria better, and add a few more to what we consider errors worth stopping on
3. Add more criteria to Pints until we agree more closely on when to halt.

Option 3 seems the best to me - it would also benefit other optimisers, and makes Pints a more predictable tool (I hate the idea of having a method you call that changes the rules of how you interact, which is what 1 amounts too). It's a bit of work though...

**Comment by MichaelClerx** [comment]
My first guess of what to implement would be the `tolx` criterion: Has `x` changed by more than Y over the last N iterations? Open questions are which `x` (the mean of the particles? the xbest solution?) and do we take scaling (cma-es sigma) into account.

**Comment by mirams** [comment]
Definitely good to have a concept of stopping on (probably best) parameter change as well as (best) objective change...

**Comment by chonlei** [comment]
I think their `tolf` criterion, tolerance for the objective function to be minimised, is very important too! We know CMA-ES (at least my feeling of it) is very good at finding global minimum but quite bad at fine tunning it. So `tolf` is quite handy sometime to make it much faster if we don't really care about that last `1e-8` digits that it spends hours to fine tune...

**Comment by MichaelClerx** [comment]
We have that! https://pints.readthedocs.io/en/latest/optimisers/running.html#pints.Optimisation.set_threshold

**Comment by mirams** [comment]
Our docs sound like stopping when objective hits a given threshold (absolute value f(x)<=tol) - is that right? 

Whereas most optimisers have an option to stop when they think they are within a given distance of the best they are likely to find (relative so it thinks current optimum <= best optimum + tol)

**Comment by MichaelClerx** [comment]
Our default criteria are maximum number of iterations and maximum unchanged iterations. The objective threshold is disabled by default: https://pints.readthedocs.io/en/latest/optimisers/running.html#pints.Optimisation.set_max_iterations

I guess you were looking at a different part of the docs? We should make this more clear then!

**Comment by mirams** [comment]
I was looking at this: https://pints.readthedocs.io/en/latest/optimisers/running.html#pints.Optimisation.set_threshold which isn't the same (or isn't described the same) as a matlab/cma-es `TolFun` :
![image](https://user-images.githubusercontent.com/858776/52342908-976e2e80-2a0e-11e9-9341-308545255eea.png)

**Comment by MichaelClerx** [comment]
I'm not sure what you're asking anymore :D

CMA-ES uses these criteria internally (not all always enabled):
```
# tolconditioncov(1e14) ERROR CHECK: Condition of covariance matrix
# tolfacupx(1e3) ERROR CHECK: Massive increases in step-size
# timeout(inf) Limits real time
# tolupsigma(1e20) Creeping / slow improvement
# tolstagnation(has default) Threshold in unchanged iterations
# tolx(1e-11) Threshold on change in position
# ftarget(no default) Threshold on target function value
# tolfun(1e-11) Threshold on change in target value (one iteration)
# tolfunhist(1e-12) Threshold on long-term change in target value
# maxfevals(inf) Max function evaluations
# maxiter(has default) Max iterations
# noeffectcoord ?
# noeffectaxis ?
# flat fitness: CMAES thinks the landscape is flat
# "||xmean||^2<ftarget"
# callback: User callback triggered stop
```
But we overrule those in the Optimisation class, to make all optimisers work the same

Our default criteria are
1. `TolFun` in the above matlab diagram, where we stop after N iterations with less than X absolute change in the objective. This is  "I haven't improved more than tol for a while so I'm not likely to"
2. Max iterations

Optionally you can add
3. A threshold below (or above for loglikelihoods) which the optimization stops

I'm happy to (add tickets to) add more stopping criteria?

**Comment by mirams** [comment]
I'm saying that if `Optimisation.set_threshold()` is setting `tolf` or `TolFun` then that isn't what our documentation says!

**Comment by MichaelClerx** [comment]
Sorry, I'll try that again

> Our docs sound like stopping when objective hits a given threshold (absolute value f(x)<=tol) - is that right?

Yes!

:D

**Comment by mirams** [comment]
> We have that! https://pints.readthedocs.io/en/latest/optimisers/running.html#pints.Optimisation.set_threshold

OK - so we **don't** have that (Chon was asking about `tolf`)! 

We should probably make a common stopping interface to do `tolf` and `tolx` methods on lots of optimisers. They are both really useful, and most optimisers use them.

(Although `tolf` might actually want to set CMA-ES's `tolfunhist` rather than `tolfun`... or implement something similar to `tolfunhist` if we are going to implement the checks ourselves).

**Comment by mirams** [comment]
(we seem to be sort of setting a `tolx` in the `set_max_unchanged...` last argument, but it probably wants to be a separate method and have the chance to be a vector).

**Comment by MichaelClerx** [comment]
Ah. Ok now I get it!

So #643 should fix this?

**Comment by mirams** [comment]
Yeah, and we should consider making something that has `tolf` or `TolFun` in the name, or at least in the documentation so it comes up in a search, as I don't think people will necessarily find the current way of setting it, even though it is a good way to think about it.

**Comment by mirams** [comment]
Setting a "threshold" on the optimiser I expected to implement `tolf` - as we won't normally know what the minimum objective function value we are looking for is!

**Comment by mirams** [comment]
Looking at parallels to Matlab's optimisation options: https://uk.mathworks.com/help/matlab/ref/optimset.html and SciPy optimise options: https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.show_options.html + similar should tell us what is sensible to include.
